### PR TITLE
Teach config bootstrapper to delete stale keys

### PR DESCRIPTION
**Pull request by listx**
The config-bootstrapper works by looking at local files, and treating
all files it finds as a file addition. Then these files are converted
into "ConfigMapUpdate" entries, and sent to updateconfig.Update() to
update the corresponding ConfigMap key.

The problem is that the above mechanism doesn't work for deleting keys.
For example, deleting a file and then running config-bootstrapper again
will simply "update" all other files except the deleted one (because it
doesn't see it in the local filesystem).

Teach updateconfig.Update() how to delete ConfigMap keys that are not
updated. Do this by calling updateconfig.MarkStaleKeysForDeletion()
which checks for those ConfigMap keys that were left untouched and
creates corresponding ConfigMapUpdate entries signaling their deletion.

An alternative way would be to call the "patch" verb to delete those
keys individually. However, this would incur additional API overhead of
needing to call both "update" (to send the locally modified ConfigMap
object) and the "patch" verb (to delete untouched keys), not to mention
that we would be bypassing the existing mechanism we already have in
updateconfig.Update(), which already knows how to delete keys inside the
locally modified ConfigMap object [1]).

For these reasons, continue to use the "Update" verb without "Patch". We
have to send the entire object over (instead of only sending a diff),
but we've already been doing this and we don't want to make large
changes in this sensitive area. The previous commit records for the
record why using "Update" isn't very accurate in terms of unit tests,
but this is a minor cost to pay in exchange for making this commit much
simpler (no need to modify how deletions are done at [1]).

[1] https://github.com/kubernetes/test-infra/blob/d3ef1505f12cd2dd9f2bfc898bb987dc22c48066/prow/plugins/updateconfig/updateconfig.go#L140-L144

### CODE REVIEW

**Comment by listx** [comment]
/hold

In the first commit of this PR, I tried to add a failing test. But for some reason that test passes even without the necessary improvements introduced in the second commit. I have a guess as to why this is, but I'm not sure.

Before merging this I'll delete that first commit (or, if reviewers can suggest a way to make that test fail, improve it so that it fails before the second commit, so that it only passes with the improvements in the second commit). For now I'm leaving it here for the record.

**Comment by airbornepony** [comment]
In the description, there's a typo: s/thsoe/those/. Will review the code more closely during the day.

**Comment by listx** [comment]
> In the first commit of this PR, I tried to add a failing test. But for some reason that test passes even without the necessary improvements introduced in the second commit. I have a guess as to why this is, but I'm not sure.

I will check to confirm that the fake ConfigMap interface we have doesn't support the K8s API method we are using (likely PATCH). Making the fake smarter should fix this problem.

**Comment by listx** [comment]
It looks like the fake ConfigMap interface [we use](https://github.com/kubernetes/test-infra/blob/d3ef1505f12cd2dd9f2bfc898bb987dc22c48066/prow/cmd/config-bootstrapper/main_test.go#L457) is [not something we wrote ourselves](https://pkg.go.dev/k8s.io/client-go/kubernetes/fake#NewSimpleClientset). Looking at that library, it uses the [NewObjectTracker](https://pkg.go.dev/k8s.io/client-go/testing#NewObjectTracker) which implements the [Update method](https://github.com/kubernetes/client-go/blob/b0ed45df66138ec7761c6db922d8f96672773dd5/testing/fixture.go#L359) by doing a DeepCopy and [replacing the old object with a fresh copy of the new object](https://github.com/kubernetes/client-go/blob/b0ed45df66138ec7761c6db922d8f96672773dd5/testing/fixture.go#L411).

While it appears that the fake does support [Patch operations](https://github.com/kubernetes/client-go/blob/b0ed45df66138ec7761c6db922d8f96672773dd5/testing/fixture.go#L145), we currently only use the [Update operation](https://github.com/kubernetes/test-infra/blob/d3ef1505f12cd2dd9f2bfc898bb987dc22c48066/prow/plugins/updateconfig/updateconfig.go#L184) in the ConfigUpdater struct. I don't want to change how that struct works because although I've had a glance over https://kubernetes.io/docs/reference/using-api/api-concepts/#patch-and-apply I don't understand all of the nuances of using Patch over Update.

So for now I will just leave the test case as is.

**Comment by listx** [comment]
Instructions for reviewers: please read the first commit's description for my current understanding of this area. Basically, I don't want to overhaul how the the ConfigUpdater works, so I chose to leave the "Update" vs "Patch" debate alone. This is reiterated in the second commit's description.

That being said, I believe the higher-level tests in the 2nd commit should at least demonstrate that the new helper functions I've added do the right thing, so please take a close look there. Thanks.

File: prow/cmd/config-bootstrapper/main.go
Comment (listx) [reply prow/cmd/config-bootstrapper/main.go]: Done.

File: prow/plugins/updateconfig/updateconfig.go
Comment (listx) [reply prow/plugins/updateconfig/updateconfig.go]: Done. We don't destructively modify the configMap parameter any more.

**Review by cjwagner — APPROVED** [review]

**Review by airbornepony — APPROVED** [review]

**Pull request merged**
